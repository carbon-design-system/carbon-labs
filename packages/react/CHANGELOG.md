# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

# 0.1.0 (2025-01-30)


### Bug Fixes

* get web-component test for example-button working ([#357](https://github.com/carbon-design-system/carbon-labs/issues/357)) ([5ae7525](https://github.com/carbon-design-system/carbon-labs/commit/5ae7525867850c2e2bc6fd80834b054759437412))


### Features

* script to generate base component in react + web-components ([#354](https://github.com/carbon-design-system/carbon-labs/issues/354)) ([3c24d37](https://github.com/carbon-design-system/carbon-labs/commit/3c24d375f2b8773f7c4dd704197ae6267d919685))
* **storybook:** add root level storybook instance for landing page ([#358](https://github.com/carbon-design-system/carbon-labs/issues/358)) ([473596b](https://github.com/carbon-design-system/carbon-labs/commit/473596b99b5aef9c1dbe261345d24b3033dd56e2))
* **TextHighlighter:** Add TextHighlighter component - React ([#384](https://github.com/carbon-design-system/carbon-labs/issues/384)) ([61dca6b](https://github.com/carbon-design-system/carbon-labs/commit/61dca6ba71aafcab4a101cb21fc898a630c4c333))
