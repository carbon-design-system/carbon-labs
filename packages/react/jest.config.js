/**
 * @license
 *
 * Copyright IBM Corp. 2024
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

export default {
  testEnvironment: 'jsdom',
  preset: 'ts-jest',
  moduleNameMapper: {
    '\\.(css|less|scss|sass)$': '<rootDir>/__mocks__/styleMock.js',
  },
  moduleFileExtensions: ['ts', 'tsx', 'js', 'jsx', 'json', 'node'],
  transform: {
    '^.+\\.(ts|tsx)$': 'ts-jest',
    '^.+\\.(js|jsx)$': ['babel-jest', { configFile: './.babelrc' }],
  },
  transformIgnorePatterns: ['/node_modules/(?!@carbon-labs)'],
  setupFilesAfterEnv: ['<rootDir>/jest.setup.ts'],

  // Test file patterns to match
  testMatch: [
    '<rootDir>/src/components/**/__tests__/**/*.[jt]s?(x)',
    '<rootDir>/src/**/?(*.)+(spec|test).[jt]s?(x)',
  ],
};
