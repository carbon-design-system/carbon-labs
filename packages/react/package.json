{
  "name": "@carbon/labs-react",
  "version": "0.0.0",
  "private": true,
  "type": "module",
  "description": "React components for Carbon Labs",
  "license": "Apache-2.0",
  "repository": "https://github.com/carbon-design-system/carbon-labs",
  "scripts": {
    "storybook": "storybook dev -p 3000",
    "storybook:build": "storybook build",
    "test": "jest",
    "test:updateSnapshot": "jest -u"
  },
  "files": [
    "es",
    "lib"
  ],
  "peerDependencies": {
    "react": "^16.8.6 || ^17.0.1 || ^18.2.0",
    "react-dom": "^16.8.6 || ^17.0.1 || ^18.2.0"
  },
  "dependencies": {
    "@carbon/react": "^1.71.1",
    "@carbon/styles": "^1.70.0",
    "react": "^18.3.1",
    "react-dom": "^18.3.1"
  },
  "devDependencies": {
    "@babel/core": "^7.26.0",
    "@babel/plugin-proposal-export-default-from": "^7.25.9",
    "@babel/plugin-transform-class-properties": "^7.25.9",
    "@babel/plugin-transform-export-namespace-from": "^7.25.9",
    "@babel/plugin-transform-react-constant-elements": "^7.25.9",
    "@babel/preset-env": "^7.26.0",
    "@babel/preset-react": "^7.26.3",
    "@babel/preset-typescript": "^7.26.0",
    "@mdx-js/react": "^3.1.0",
    "@rollup/plugin-babel": "^6.0.0",
    "@rollup/plugin-commonjs": "^28.0.0",
    "@rollup/plugin-node-resolve": "^16.0.0",
    "@rollup/plugin-typescript": "^12.1.1",
    "@storybook/addon-docs": "^8.4.7",
    "@storybook/addon-essentials": "^8.4.7",
    "@storybook/addon-interactions": "^8.4.7",
    "@storybook/addon-onboarding": "^8.4.7",
    "@storybook/addon-storysource": "^8.4.7",
    "@storybook/addon-webpack5-compiler-babel": "^3.0.3",
    "@storybook/blocks": "^8.4.7",
    "@storybook/react": "^8.4.7",
    "@storybook/react-webpack5": "^8.4.7",
    "@storybook/test": "^8.4.7",
    "@testing-library/jest-dom": "^6.0.0",
    "@testing-library/react": "^16.1.0",
    "@testing-library/user-event": "^14.5.2",
    "@types/babel__core": "^7",
    "@types/babel__preset-env": "^7",
    "@types/jest": "^29.0.0",
    "@types/node": "^16.18.121",
    "@types/prop-types": "^15",
    "@types/react": "^18.3.12",
    "@types/react-dom": "^18.3.1",
    "autoprefixer": "^10.4.20",
    "babel-jest": "^29.7.0",
    "babel-plugin-dev-expression": "^0.2.3",
    "browserslist-config-carbon": "^11.2.0",
    "css-loader": "^7.1.2",
    "jest": "^29.7.0",
    "jest-environment-jsdom": "^29.7.0",
    "mini-css-extract-plugin": "^2.9.2",
    "postcss": "^8.4.49",
    "postcss-loader": "^8.1.1",
    "prop-types": "^15.8.1",
    "remark-gfm": "^4.0.0",
    "rollup-plugin-copy": "^3.5.0",
    "rollup-plugin-strip-banner": "^3.0.0",
    "sass-loader": "^16.0.4",
    "storybook": "^8.4.7",
    "storybook-addon-accessibility-checker": "^3.1.61-rc.3",
    "style-loader": "^4.0.0",
    "ts-jest": "^29.2.5",
    "typescript": "^5.2.2",
    "typescript-config-carbon": "^0.3.0",
    "webpack": "^5.97.0"
  }
}
