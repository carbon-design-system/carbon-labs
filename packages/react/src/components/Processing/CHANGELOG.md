# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

# [0.4.0](https://github.com/carbon-design-system/carbon-labs/compare/@carbon-labs/react-processing@0.3.0...@carbon-labs/react-processing@0.4.0) (2025-05-28)


### Bug Fixes

* update exports to match UI shell ([#605](https://github.com/carbon-design-system/carbon-labs/issues/605)) ([1bf4682](https://github.com/carbon-design-system/carbon-labs/commit/1bf46822620b44cc1ad9ce58913bf26e9c3a2ca1))





# [0.3.0](https://github.com/carbon-design-system/carbon-labs/compare/@carbon-labs/react-processing@0.2.0...@carbon-labs/react-processing@0.3.0) (2025-04-04)

**Note:** Version bump only for package @carbon-labs/react-processing





# [0.2.0](https://github.com/carbon-design-system/carbon-labs/compare/@carbon-labs/react-processing@0.1.0...@carbon-labs/react-processing@0.2.0) (2025-03-26)


### Bug Fixes

* **processing:** remove button import + fix storybook style ([#534](https://github.com/carbon-design-system/carbon-labs/issues/534)) ([b1d514c](https://github.com/carbon-design-system/carbon-labs/commit/b1d514c1181eeb4442848e5a036b0db27f9b3899))





# 0.1.0 (2025-03-26)


### Features

* **processing:** add new processing component ([#490](https://github.com/carbon-design-system/carbon-labs/issues/490)) ([18bd75c](https://github.com/carbon-design-system/carbon-labs/commit/18bd75c0751eae30bab9d965715302c05213234c))
