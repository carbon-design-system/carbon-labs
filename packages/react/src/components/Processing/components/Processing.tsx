/**
 * @license
 *
 * Copyright IBM Corp. 2025
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
import PropTypes from 'prop-types';
import React from 'react';
import { usePrefix } from '@carbon-labs/utilities/es/index.js';

/** Primary UI component for user interaction */

export interface ProcessingProps {
  loop?: boolean;
}

const Processing: React.FC<ProcessingProps> = ({
  loop = true,
  ...rest
}: ProcessingProps) => {
  const prefix = usePrefix();
  const blockClass = `${prefix}--processing`;

  const getAnimationEffect = () => {
    if (loop == true) {
      return 'linear';
    } else {
      return 'linear--no-loop';
    }
  };

  return (
    <div className={`${blockClass} ${blockClass}__${getAnimationEffect()}`}>
      <svg className={`${blockClass}__dots`} viewBox="0 0 32 32">
        <circle
          className={`${blockClass}__dot ${blockClass}__dot--left`}
          cx="8"
          cy="16"
        />
        <circle
          className={`${blockClass}__dot ${blockClass}__dot--center`}
          cx="16"
          cy="16"
          r="2"
        />
        <circle
          className={`${blockClass}__dot ${blockClass}__dot--right`}
          cx="24"
          cy="16"
          r="2"
        />
      </svg>
    </div>
  );
};

(Processing as React.FC).displayName = 'Processing';
(Processing as React.FC).propTypes = {
  /**
   * Specify whether the animation should loop
   */
  loop: PropTypes.bool,
};

export default Processing;
