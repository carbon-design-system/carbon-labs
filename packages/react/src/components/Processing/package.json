{
  "name": "@carbon-labs/react-processing",
  "version": "0.4.0",
  "publishConfig": {
    "access": "public",
    "provenance": true
  },
  "type": "module",
  "description": "Carbon Labs - Processing",
  "license": "Apache-2.0",
  "repository": {
    "type": "git",
    "url": "https://github.com/carbon-design-system/carbon-labs",
    "directory": "src/components/Processing"
  },
  "types": "./es/index.d.ts",
  "exports": {
    ".": {
      "import": "./es/index.js",
      "require": "./lib/index.js"
    },
    "./es/*": "./es/*",
    "./lib/*": "./lib/*",
    "./scss/*": "./scss/*"
  },
  "files": [
    "es",
    "lib",
    "scss",
    "telemetry.yml"
  ],
  "scripts": {
    "build": "node ../../../tasks/build.js",
    "clean": "rm -rf {es,lib,scss}",
    "postinstall": "ibmtelemetry --config=telemetry.yml",
    "telemetry:config": "npx -y @ibm/telemetry-js-config-generator generate --id 399aab88-efa6-421f-91f7-0be428e0b6c2 --endpoint https://www-api.ibm.com/ibm-telemetry/v1/metrics --files ./components/**/*.(tsx|js|jsx)"
  },
  "devDependencies": {
    "@carbon-labs/utilities": "canary"
  },
  "dependencies": {
    "@ibm/telemetry-js": "^1.9.1"
  }
}
