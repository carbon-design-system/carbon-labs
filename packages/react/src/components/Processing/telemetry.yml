# yaml-language-server: $schema=https://unpkg.com/@ibm/telemetry-config-schema@v1/dist/config.schema.json

version: 1
projectId: 399aab88-efa6-421f-91f7-0be428e0b6c2
endpoint: https://www-api.ibm.com/ibm-telemetry/v1/metrics
collect:
  jsx:
    elements:
      allowedAttributeNames:
        # Processing
        - loop
        # React
        - key
        - ref
  npm:
    dependencies: null
  js:
    functions: {}
    tokens: null
