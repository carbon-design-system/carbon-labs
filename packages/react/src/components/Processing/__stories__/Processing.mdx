import { ArgTypes, Canvas, Meta } from '@storybook/blocks';
import * as ProcessingStories from './Processing.stories';
import Processing from '../components/Processing';
import AIChatLoop from './images/ai_chat_loop.gif';
import NoLoopLight from './images/processing_element_load_unload_light_128x128.gif';
import NoLoopDark from './images/processing_element_load_unload_dark_128x128.gif';
import LoopLight from './images/processing_element_loop_light_128x128.gif';
import LoopDark from './images/processing_element_loop_dark_128x128.gif';

<Meta isTemplate />

# Processing

- **Initiative owner(s):** Drew Glapa & Krista Starr
- **Status:** Draft
- **Target library:** TBD
- **Target library maintainer(s) / PR Reviewer(s):** N/A
- **Support channel:** `#carbon-labs`

{/* <!-- START doctoc generated TOC please keep comment here to allow auto update --> */}
{/* <!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE --> */}

> ðŸ’¡ Check our
> [Stackblitz](https://stackblitz.com/github/carbon-design-system/carbon-labs/tree/main/examples/react/Processing)
> example implementation.

[![Edit carbon-labs](https://developer.stackblitz.com/img/open_in_stackblitz.svg)](https://stackblitz.com/github/carbon-design-system/carbon-labs/tree/main/examples/react/Processing)

## Table of Contents

- [Overview](#overview)
- [Getting started](#getting-started)

{/* <!-- END doctoc generated TOC please keep comment here to allow auto update --> */}

## Overview

Processing element is used to indicate that data is currently being
processed/loaded and the user should wait.

<Canvas
  of={ProcessingStories.LinearLoop}
  additionalActions={[
    {
      title: 'Open in Stackblitz',
      onClick: () => stackblitzPrefillConfig(ProcessingStories.LinearLoop),
    },
  ]}
/>

<img src={AIChatLoop} alt="AI chat window example" width={319} />

**Ex.** Content processing in a chat shell.

### Using Processing element in Figma

As a designer, you can use the Processing component in Figma to create a more
interactive prototype. This can be done by downloading the desired `.gif` below
and adding to Figma.

|             |                                     Light mode (white, g10)                                     |                                    Dark mode (g90, g100)                                    |
| :---------- | :---------------------------------------------------------------------------------------------: | :-----------------------------------------------------------------------------------------: |
| **No Loop** | <div style={{ background: '#ffffff' }}><img src={NoLoopLight} alt="No loop light mode" /></div> | <div style={{background: '#161616'}}><img src={NoLoopDark} alt="No loop dark mode" /></div> |
| **Loop**    |   <div style={{ background: '#ffffff' }}><img src={LoopLight} alt="Loop light mode" /></div>    |  <div style={{ background: '#161616' }}><img src={LoopDark} alt="Loop dark mode" /></div>   |

## Getting started

Here's a quick example to get you started.

```bash
yarn add @carbon/react
yarn add @carbon-labs/react-processing
```

### JS (via import)

```javascript
import { Processing } from '@carbon-labs/react-processing/es/index';

function App() {
  return <Processing />;
}

export default App;
```

### SCSS

In your styles file import

```
@use '@carbon/react';
@use '@carbon-labs/react-processing/scss/processing';
```

## Component API

<ArgTypes />

## Example usages

Example of Processing element w/ loop

```jsx
<Processing loop />
```
