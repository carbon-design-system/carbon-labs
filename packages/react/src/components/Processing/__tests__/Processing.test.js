/**
 * @license
 *
 * Copyright IBM Corp. 2025
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import { render } from '@testing-library/react';
import React from 'react';
import '@testing-library/jest-dom';

import Processing from '../components/Processing';
jest.mock('./processing.scss', () => ({}));
describe('Processing', () => {
  describe('renders as expected - Component API', () => {
    it('should match snapshot', () => {
      const { container } = render(<Processing />);
      expect(container).toMatchSnapshot();
    });
  });
});
