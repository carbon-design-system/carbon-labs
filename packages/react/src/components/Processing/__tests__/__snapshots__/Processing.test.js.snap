// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Processing renders as expected - Component API should match snapshot 1`] = `
<div>
  <div
    class="clabs--processing clabs--processing__linear"
  >
    <svg
      class="clabs--processing__dots"
      viewBox="0 0 32 32"
    >
      <circle
        class="clabs--processing__dot clabs--processing__dot--left"
        cx="8"
        cy="16"
      />
      <circle
        class="clabs--processing__dot clabs--processing__dot--center"
        cx="16"
        cy="16"
        r="2"
      />
      <circle
        class="clabs--processing__dot clabs--processing__dot--right"
        cx="24"
        cy="16"
        r="2"
      />
    </svg>
  </div>
</div>
`;
